Update admission control docs for v23.2 (#17999)
cockroachdb/docs: src/current/v23.2/admission-control.md

CockroachDB supports an admission control system to maintain cluster performance and availability when some nodes experience high load. When admission control is enabled, CockroachDB sorts request and response operations into work queues by priority, giving preference to higher priority operations. Internal operations critical to node health, like [node liveness heartbeats]({% link {{ page.version.version }}/cluster-setup-troubleshooting.md %}#node-liveness-issues), are high priority. The admission control system also prioritizes transactions that hold [locks]({% link {{ page.version.version }}/crdb-internal.md %}#cluster_locks), to reduce [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention) and release locks earlier.

> [XXX](): check in with #cc-serverless team on the conflicting Serverless statements in here.

Reviewer: This is incorrect, since we look at AC behavior when investigating issues in serverless clusters.
@JeffSwenson

Author: thanks, I have removed the note that said "Admission control is not available for CockroachDB serverless clusters"

For storage IO, the goal is to prevent the [storage layer's log-structured merge tree]({% link {{ page.version.version }}/architecture/storage-layer.md %}#log-structured-merge-trees) (LSM) from experiencing high [read amplification]({% link {{ page.version.version }}/architecture/storage-layer.md %}#read-amplification), which slows down reads, while also maintaining the ability to absorb bursts of writes.

Admission control works on a per-[node]({% link {{ page.version.version }}/architecture/overview.md %}#node) basis, since even though a large CockroachDB cluster may be well-provisioned as a whole, individual nodes are stateful and may experience performance [hot spots]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#hot-spots).

For more details about how the admission control system works, see:

- The [Admission Control tech note](https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/admission_control.md).
- The blog post [Here's how CockroachDB keeps your database from collapsing under load](https://www.cockroachlabs.com/blog/admission-control-in-cockroachdb/).

> 

Reviewer: And https://www.cockroachlabs.com/blog/rubbing-control-theory/ which talks about CPU tokens for elastic work.

Author: added it! thanks

CockroachDB supports an admission control system to maintain cluster performance and availability when some nodes experience high load. When admission control is enabled, CockroachDB sorts request and response operations into work queues by priority, giving preference to higher priority operations. Internal operations critical to node health, like [node liveness heartbeats]({% link {{ page.version.version }}/cluster-setup-troubleshooting.md %}#node-liveness-issues), are high priority. The admission control system also prioritizes transactions that hold [locks]({% link {{ page.version.version }}/crdb-internal.md %}#cluster_locks), to reduce [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention) and release locks earlier.

## How admission control works

At a high level, the admission control system works by queueing requests to use the following system resources:

- CPU
- Storage IO (writes to disk)

> For CPU, different types of usage are queued differently based on priority to allow important work to make progress even under high CPU utilization.

Reviewer 2: Not sure if worth linking "high CPU" to: https://www.cockroachlabs.com/docs/stable/common-issues-to-monitor#cpu-usage

Up to you!

Author: Added the link! Always love to have more links!

- The node has high CPU usage, visible in the **CPU percent** graph in the [**Overload** dashboard]({% link {{ page.version.version }}/ui-overload-dashboard.md %}#cpu-percent).
- The node is experiencing out-of-memory errors, visible in the **Memory Usage** graph in the [**Hardware** dashboard]({% link {{ page.version.version }}/ui-hardware-dashboard.md %}#memory-usage). Even though admission control does not explicitly target controlling memory usage, it can reduce memory usage as a side effect of delaying the start of operation execution when the CPU is overloaded.

> ## Operations subject to Admission Control

Reviewer 2: ```suggestion
## Operations subject to admission control
```

Author: Fixed!

> Almost all database operations that use CPU or perform Storage IO are controlled by the admission control system. From a user's perspective, specific operations that are affected by admission control include:

Reviewer 2: ```suggestion
Almost all database operations that use CPU or perform storage IO are controlled by the admission control system. From a user's perspective, specific operations that are affected by admission control include:
```

Author: Fixed!

- [Changefeeds]({% link {{ page.version.version }}/create-and-configure-changefeeds.md %}).
- [Intent resolution]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#write-intents).

The following operations are not subject to admission control:

- SQL writes are not subject to admission control on [follower replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft) by default, unless those writes occur in transactions that are subject to a Quality of Service (QoS) level as described in [Set quality of service level for a session](#set-quality-of-service-level-for-a-session). In order for writes on follower replicas to be subject to admission control, the setting `default_transaction_quality_of_service=background` must be used.

{{site.data.alerts.callout_info}}
Admission control is beneficial when overall cluster health is good but some nodes are experiencing overload. If you see these overload scenarios on many nodes in the cluster, that typically means the cluster needs more resources.
{{site.data.alerts.end}}

## Enable and disable admission control

Admission control is enabled by default.

> 

Reviewer 2: Extra line break here.

Author: ya I added that for emphasis but I'm not wedded to it.  Removed

To enable or disable admission control, use the following [cluster settings]({% link {{ page.version.version }}/cluster-settings.md %}):

- `admission.kv.enabled` for work performed by the [KV layer]({% link {{ page.version.version }}/architecture/distribution-layer.md %}).
- `admission.sql_kv_response.enabled` for work performed in the SQL layer when receiving [KV responses]({% link {{ page.version.version }}/architecture/distribution-layer.md %}).
- `admission.sql_sql_response.enabled` for work performed in the SQL layer when receiving [DistSQL responses]({% link {{ page.version.version }}/architecture/sql-layer.md %}#distsql).

> When you enable or disable admission control settings for one layer, we recommend that you enable or disable them for **all layers**.

Reviewer 2: I know, I know

```suggestion
When you enable or disable admission control settings for one layer, Cockroach Labs recommends that you enable or disable them for **all layers**.
```

Author: lol

worst ever addition to the CRL style guide IMO.  but I will adhere to the letter of the law of the land

Author: (PS fixed it)

## Operations subject to Admission Control

Almost all database operations that use CPU or perform Storage IO are controlled by the admission control system. From a user's perspective, specific operations that are affected by admission control include:

- [XXX](): ADD LINKS BACK AND FORTH TO/FROM THINGS
- [XXX]() ASK SUMEER ABOUT WHETHER THIS LIST OF THINGS MAKES SENSE
- [General SQL queries]({% link {{ page.version.version }}/selection-queries.md %}) have their CPU usage subject to admission control, as well as storage IO for writes to [leaseholder replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#leases).
- [Bulk data imports]({% link {{ page.version.version }}/import-into.md %}).
- [Backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}).
- [Schema changes]({% link {{ page.version.version }}/online-schema-changes.md %}), including index and column backfills (on both the [leaseholder replica]({% link {{ page.version.version }}/architecture/replication-layer.md %}#leases) and [follower replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft)).

> - [Deletes]({% link {{ page.version.version }}/delete-data.md %}) (including those initiated by [row-level TTL jobs]({% link {{ page.version.version }}/row-level-ttl.md %}))

Reviewer: Also scans done by TTL are subject to CPU admission control

Author: I have added a clause to this line to note that info, PTAL and let me know if it seems reasonable

> - [Follower replication work]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft) (including the sending of [snapshots]({% link {{ page.version.version }}/architecture/replication-layer.md %}#snapshots)).

Reviewer: Snapshot receipts are not subject to admission control. There is a more rudimentary legacy throttling mechanism there.

Author: I have deleted the parenthetical mentioning snapshots, PTAL

- [Raft log entries being written to disk]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft).
- [Changefeeds]({% link {{ page.version.version }}/create-and-configure-changefeeds.md %}).
- [Intent resolution]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#write-intents).

The following operations are not subject to admission control:

- [XXX]() ASK SUMEER ABOUT WHAT THINGS CAN/SHOULD GO HERE
- SQL writes are not subject to admission control on [follower replicas]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft) by default, unless those writes occur in transactions that are subject to a Quality of Service (QoS) level as described in [Set quality of service level for a session](#set-quality-of-service-level-for-a-session).

> 

Reviewer: specifically it needs to be `default_transaction_quality_of_service=background` for it to be subject to admission control on the follower.

Organizing operations by priority can mean that higher priority operations consume all the available resources while lower priority operations remain in the queue until the operation times out.

## Considerations

> Too many connections per [gateway node]({% link {{ page.version.version }}/architecture/sql-layer.md %}#gateway-node) can also lead to cluster overload. [Client connections]({% link {{ page.version.version }}/connection-parameters.md %}) are not managed by the admission control subsystem. {% include {{page.version.version}}/sql/server-side-connection-limit.md %}

Reviewer 2: This may be a bit clearer

```suggestion
[Client connections]({% link {{ page.version.version }}/connection-parameters.md %}) are not managed by the admission control subsystem. Too many connections per [gateway node]({% link {{ page.version.version }}/architecture/sql-layer.md %}#gateway-node) can also lead to cluster overload. 

{% include {{page.version.version}}/sql/server-side-connection-limit.md %}
```

Author: love this edit - much clearer, thank you!
